cf-check/backup.c

```diff
@@ -0,0 +1,217 @@
+#include <platform.h>
+#include <backup.h>
+
+#if defined (__MINGW32__)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup not supported on Windows\n");
+    return 1;
+}
+
+#elif ! defined (LMDB)
+
+int backup_main(int argc, char **argv)
+{
+    printf("backup only implemented for LMDB.\n");
+    return 1;
+}
+
+#else
+
+#include <stdio.h>
+
+bool copy_file(const char *src, const char *dst)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+
+    printf("Copying: '%s' -> '%s'\n", src, dst);
+    FILE *in = fopen(src, "r");
+    if (in == NULL)
+    {
+        printf("Could not open '%s' (%s)\n", src, strerror(errno));
+        return false;
+    }
+    FILE *out = fopen(dst, "w");
+    if (out == NULL)
+    {
+        printf("Could not open '%s' (%s)\n", dst, strerror(errno));
+        fclose(in);
+        return false;
+    }
+    size_t read = 0;
+    size_t written = 0;
+    bool ret = true;
+    do
+    {
+        const size_t bufsize = 1024;
+        char buf[1024] = { 0 };
+
+        read = fread(buf, sizeof(char), bufsize, in);
+
+        written = fwrite(buf, sizeof(char), read, out);
+        while ((!ferror(out)) && (written < read))
+        {
+            written += fwrite(buf + written, sizeof(char), read - written, out);
+        }
+
+        if (read != written || ferror(in) || ferror(out))
+        {
+            break;
+        }
+        if (feof(in))
+        {
+            read = 0;
+        }
+    } while (read > 0);
+
+    if (ferror(in))
+    {
+        printf("Error encountered while reading '%s'\n", src);
+        ret = false;
+    }
+    else if (ferror(out))
+    {
+        printf("Error encountered while writing '%s'\n", dst);
+        ret = false;
+    }
+    else if (read != 0 && read != written)
+    {
+        printf("Did not copy the whole file\n");
+        ret = false;
+    }
+
+    const int i = fclose(in);
+    if (i != 0)
+    {
+        printf("Error encountered while closing '%s' (%s)\n", src, strerror(errno));
+        ret = false;
+    }
+    const int o = fclose(out);
+    if (o != 0)
+    {
+        printf("Error encountered while closing '%s' (%s)\n", dst, strerror(errno));
+        ret = false;
+    }
+    return ret;
+}
+
+const char *filename_part(const char *path)
```

Comment:
There is a `basename()` function in *string.h*, why not use that one?

Reply:
Because it modifies its argument.